libm/src/math/arch/i586.rs

```diff
@@ -1,37 +1,71 @@
+//! main implementation fails with the x87 FPU used by
+//! debian i386, probably due to excess precision issues.
+//!
+//! See https://github.com/rust-lang/compiler-builtins/pull/976 for discussion on why these
+//! functions are implemented in this way.
 
-use super::super::fabs;
+// FIXME: when the MSRV allows, use naked functions instead.
 
-/// Use an alternative implementation on x86, because the
-/// main implementation fails with the x87 FPU used by
-/// debian i386, probably due to excess precision issues.
-/// Basic implementation taken from https://github.com/rust-lang/libm/issues/219.
-pub fn ceil(x: f64) -> f64 {
-    if fabs(x).to_bits() < 4503599627370496.0_f64.to_bits() {
-        let truncated = x as i64 as f64;
-        if truncated < x {
-            return truncated + 1.0;
-        } else {
-            return truncated;
-        }
-    } else {
-        return x;
+pub extern "C" fn ceil(mut x: f64) -> f64 {
+    // We save and later restore the FPU control word.
+    let mut cw_stash = core::mem::MaybeUninit::<u16>::uninit();
+    let mut cw_tmp = core::mem::MaybeUninit::<u16>::uninit();
+    unsafe {
+        core::arch::asm!(
+            "fstcw   ({stash_ptr})",      // Save the cw
+            "movw    ({stash_ptr}), %dx", // ...
+            "orw     $0x0800, %dx",    // Set rounding control to 0b10 (+∞),
+            "andw    $0xfbff, %dx",    // preserving other controls
+            "movw    %dx, ({cw_ptr})", // Apply cw
+            "fldcw    ({cw_ptr})",     // ...
+            "fldl     ({x_ptr})",      // Push x to the stack
+            "frndint",                 // Round
+            "fldcw    ({stash_ptr})",  // Restore cw
+            "fstpl   ({x_ptr})",       // Save rounded x to mem
+            cw_ptr = in(reg) &mut cw_tmp,
+            stash_ptr = in(reg) &mut cw_stash,
+            x_ptr = in(reg) &mut x,
+            out("dx") _,               // Cw scratch
+            // All the x87 FPU stack is used, all registers must be clobbered
+            out("st(0)") _, out("st(1)") _, out("st(2)") _, out("st(3)") _,
+            out("st(4)") _, out("st(5)") _, out("st(6)") _, out("st(7)") _,
+            options(att_syntax)
+        )
     }
+    x
 }
 
-/// Use an alternative implementation on x86, because the
-/// main implementation fails with the x87 FPU used by
-/// debian i386, probably due to excess precision issues.
-/// Basic implementation taken from https://github.com/rust-lang/libm/issues/219.
-pub fn floor(x: f64) -> f64 {
-    if fabs(x).to_bits() < 4503599627370496.0_f64.to_bits() {
-        let truncated = x as i64 as f64;
-        if truncated > x {
-            return truncated - 1.0;
-        } else {
-            return truncated;
-        }
-    } else {
-        return x;
+// NOTE: this `inline(never)` is load-bearing for making functions that use it panic-free.
+// Without it `rem_pio2_large` (and any function that uses it) will contain panics.
```

Comment:
Do they actually contain panics? Or is it a false positive that the pattern used by `no_panic` doesn't handle

Reply:
I haven't been able to actually pinpoint where the reported panic really is. e.g. https://godbolt.org/z/TcEGeYos9 doesn't show any panics. So there is a chance that it's a limitation in `no_panic`.